How to handle `create_index` with small dataset

# Problem Statement

Right now, an `assert` to make the create_index call crashes, if the datasets is too small to run kmeans

```
>       self._ds.create_index(column, index_type, name, metric, kwargs)
E       pyo3_runtime.PanicException: assertion failed: data.len() > k * dimension

venv/lib/python3.11/site-packages/lance/dataset.py:611: PanicException
```

We need to handle it more gracefully.

--- a/rust/src/dataset/fragment.rs
+++ b/rust/src/dataset/fragment.rs
@@ -649,7 +649,7 @@ mod tests {
         for f in dataset.get_fragments() {
             for file in Fragment::from(f.clone()).files {
                 let p = file.path.clone();
-                paths.push(p.to_string())
+                paths.push(p);
             }
         }
 

--- a/rust/src/index/vector/kmeans.rs
+++ b/rust/src/index/vector/kmeans.rs
@@ -72,6 +72,6 @@ pub async fn train_kmeans(
         redos,
         ..Default::default()
     };
-    let model = KMeans::new_with_params(&data, dimension, k, &params).await;
+    let model = KMeans::new_with_params(&data, dimension, k, &params).await?;
     Ok(model.centroids.as_ref().clone())
 }

--- a/rust/src/utils/kmeans.rs
+++ b/rust/src/utils/kmeans.rs
@@ -146,8 +146,6 @@ async fn kmeans_random_init(
     mut rng: impl Rng,
     metric_type: MetricType,
 ) -> KMeans {
-    assert!(data.len() > k * dimension);
-
     let chosen = (0..data.len() / dimension)
         .choose_multiple(&mut rng, k)
         .to_vec();

--- a/rust/src/utils/kmeans.rs
+++ b/rust/src/utils/kmeans.rs
@@ -354,7 +354,7 @@ impl KMeans {
         if n < k {
             return Err(Error::Index {
                 message: format!(
-                    "KMeans: number of data points ({}) is smaller than k ({})",
+                    "KMeans: training does not have sufficient data points: n({}) is smaller than k({})",
                     n, k
                 ),
             });

--- a/rust/src/utils/kmeans.rs
+++ b/rust/src/utils/kmeans.rs
@@ -148,15 +148,7 @@ async fn kmeans_random_init(
     mut rng: impl Rng,
     metric_type: MetricType,
 ) -> Result<KMeans> {
-    let n = data.len() / dimension;
-    if n < k {
-        return Err(Error::Index {
-            message: format!(
-                "KMean: data length {} is less than k ({}) * dimension ({})",
-                n, k, dimension
-            ),
-        });
-    }
+    assert!(data.len() >= k * dimension);
     let chosen = (0..data.len() / dimension)
         .choose_multiple(&mut rng, k)
         .to_vec();

--- a/rust/src/utils/kmeans.rs
+++ b/rust/src/utils/kmeans.rs
@@ -506,4 +506,19 @@ impl KMeans {
 }
 
 #[cfg(test)]
-mod tests {}
+mod tests {
+    use super::*;
+
+    use arrow_array::Float32Array;
+
+    #[tokio::test]
+    async fn test_train_with_small_dataset() {
+        let data = Float32Array::from(vec![1.0, 2.0, 3.0, 4.0]);
+        match KMeans::new(&data, 2, 128, 5).await {
+            Ok(_) => assert!(false, "Should fail to train KMeans"),
+            Err(e) => {
+                assert!(e.to_string().contains("smaller than"));
+            }
+        }
+    }
+}

--- a/rust/src/utils/kmeans.rs
+++ b/rust/src/utils/kmeans.rs
@@ -507,7 +507,7 @@ mod tests {
     async fn test_train_with_small_dataset() {
         let data = Float32Array::from(vec![1.0, 2.0, 3.0, 4.0]);
         match KMeans::new(&data, 2, 128, 5).await {
-            Ok(_) => assert!(false, "Should fail to train KMeans"),
+            Ok(_) => panic!("Should fail to train KMeans"),
             Err(e) => {
                 assert!(e.to_string().contains("smaller than"));
             }

--- a/rust/src/utils/kmeans.rs
+++ b/rust/src/utils/kmeans.rs
@@ -138,14 +138,25 @@ async fn kmean_plusplus(
     kmeans
 }
 
-/// Randomly initialize kmeans centroids
+/// Randomly initialize kmeans centroids.
+///
+///
 async fn kmeans_random_init(
     data: &Float32Array,
     dimension: usize,
     k: usize,
     mut rng: impl Rng,
     metric_type: MetricType,
-) -> KMeans {
+) -> Result<KMeans> {
+    let n = data.len() / dimension;
+    if n < k {
+        return Err(Error::Index {
+            message: format!(
+                "KMean: data length {} is less than k ({}) * dimension ({})",
+                n, k, dimension
+            ),
+        });
+    }
     let chosen = (0..data.len() / dimension)
         .choose_multiple(&mut rng, k)
         .to_vec();
@@ -155,7 +166,7 @@ async fn kmeans_random_init(
     }
     let mut kmeans = KMeans::empty(k, dimension, metric_type);
     kmeans.centroids = Arc::new(builder.finish());
-    kmeans
+    Ok(kmeans)
 }
 
 pub struct KMeanMembership {
@@ -313,12 +324,17 @@ impl KMeans {
         k: usize,
         metric_type: MetricType,
         rng: impl Rng,
-    ) -> Self {
+    ) -> Result<Self> {
         kmeans_random_init(&data.data(), data.num_columns(), k, rng, metric_type).await
     }
 
     /// Train a KMeans model on data with `k` clusters.
-    pub async fn new(data: &Float32Array, dimension: usize, k: usize, max_iters: u32) -> Self {
+    pub async fn new(
+        data: &Float32Array,
+        dimension: usize,
+        k: usize,
+        max_iters: u32,
+    ) -> Result<Self> {
         let params = KMeansParams {
             max_iters,
             metric_type: MetricType::L2,
@@ -327,13 +343,23 @@ impl KMeans {
         Self::new_with_params(data, dimension, k, &params).await
     }
 
-    /// Train a KMeans model with full parameters.
+    /// Train a [`KMeans`] model with full parameters.
     pub async fn new_with_params(
         data: &Float32Array,
         dimension: usize,
         k: usize,
         params: &KMeansParams,
-    ) -> Self {
+    ) -> Result<Self> {
+        let n = data.len() / dimension;
+        if n < k {
+            return Err(Error::Index {
+                message: format!(
+                    "KMeans: number of data points ({}) is smaller than k ({})",
+                    n, k
+                ),
+            });
+        }
+
         // TODO: refactor kmeans to work with reference instead of Arc?
         let data = Arc::new(data.clone());
         let mut best_kmeans = Self::empty(k, dimension, params.metric_type);
@@ -348,7 +374,7 @@ impl KMeans {
             } else {
                 match params.init {
                     KMeanInit::Random => {
-                        Self::init_random(&mat, k, params.metric_type, rng.clone()).await
+                        Self::init_random(&mat, k, params.metric_type, rng.clone()).await?
                     }
                     KMeanInit::KMeanPlusPlus => {
                         kmean_plusplus(data.clone(), dimension, k, rng.clone(), params.metric_type)
@@ -386,7 +412,7 @@ impl KMeans {
             }
         }
 
-        best_kmeans
+        Ok(best_kmeans)
     }
 
     /// Train for one iteration.

